<div class="alert alert-block alert-success" style="font-size: 24px;">
<b>BAMS 506. Optimal Decision Making I</b> 
</div> 

### **Group Members:**

> **Saloni Annasaheb Gharge** (Student ID: 55859177)

> **Andi Wang** (Student ID: 29916046)

> **Kaveh Dorafshan** (Student ID: 91097659)

> **Vishal Kurup** (Student ID: 23966427)

<div class="alert alert-block alert-warning" style="font-size: 20px;"> 
  <b> Part I: </b>An Executive Summary  
</div>

### **Purpose and Background:**
Based on choosing the 'Option B' for our project, the purpose of this report is to provide a detailed meal plan for Starlight's "Feeding Hope" project, which aims to address malnutrition among local children due to economic hardship. The aim of the report is to design a weekly meal plan that meets the nutritional needs of children at the lowest cost, while ensuring that the diet is varied and within budget constraints.

### **Key Questions:**
Key challenges include providing balanced, nutritious meals to children within limited resources and tight budgets, while ensuring that no single food item accounts for more than 30 percent of total calorie or protein intake.

### **Analysis Methods and Approaches:**

> #### 1. How do we arrive at the optimal solution? 
> To design the optimal meal plan, we assign a coefficient to each food item and use a looping method to create formulas. These formulas multiply the nutrient content and price of each food item by its coefficient and sums them up to form the basic equation. Additionally, to ensure that "no single food item should account for more than 30% of the total calories or protein intake," we use the looping method to impose this restriction.

> #### 2. How did we come up with a one-day-a-week plan?
> We developed the one-day-a-week plan by constructing a model that applies nutritional and dietary constraints for each day of the week. Using Gurobi, we optimized the meal plan to balance nutrient intake, variety, and cost efficiency. By assigning specific constraints to each day, such as the minimum servings from food groups (e.g., vegetables, proteins, grains), we ensured that the plan provides sufficient variety and maintains a balanced diet. The model minimizes the cost by selecting affordable food combinations that meet daily nutritional goals, making the meal plan both cost-effective and nutritionally adequate for the entire week.

### **Key Findings and Suggestion:**

It is recommended to introduce more high-protein, high-calorie, low-sodium foods, such as poached eggs and beef, to increase protein and calorie intake while controlling sodium content. In addition, it is recommended to adjust food ratios to ensure balanced nutrition and develop a 7-day meal plan that provides different nutritious meals each day.

### **Conclusion:**

The implementation of this meal plan is critical to improving the nutritional status of Starlight's children. By implementing this program, children's health and school performance can be improved while providing sustainable nutrition solutions to the community. It is expected that the next step will be to implement this plan and regularly evaluate its effectiveness to ensure continuous improvement.


<div class="alert alert-block alert-warning" style="font-size: 20px;"> 
  <b> Part II: </b>Introduction  
</div>

The "Feeding Hope" project needs to simultaneously consider both dietary nutritional balance and economic costs. Specifically, the following goals have been identified:

> ### **1. Nutritional Requirements:**
>Ensure that the children’s daily food intake meets the appropriate nutritional requirements (minimum and maximum recommended daily intake for particular nutrients). The ten key nutrients include Calories (kcal), Fat (g), Sodium (mg), Carbs (g), Fiber (g), Protein (g), Vitamin A (IU), Vitamin C (mg), Calcium (mg), and Iron (mg).
The nutritionists involved in the project also recommend that to maintain a balanced diet, no single food item should account for more than 30% of the total calories or protein intake.

> ### **2. Cost Consideration:**
> The team needs to design the meal plan at the lowest possible cost.

Currently, we have thirty food options to choose from, including meats such as beef and roasted chicken, dairy products like skim milk, carbohydrates like wheat bread and spaghetti with sauce, vegetables like raw carrots and broccoli, and fruits like bananas and oranges. Each food item contains different proportions of nutrients and has a different price. 

To design the optimal meal plan, we assign a coefficient to each food item and use a looping method to create formulas. These formulas multiply the nutrient content and price of each food item by its coefficient and sums them up to form the basic equation. Additionally, to ensure that "no single food item should account for more than 30% of the total calories or protein intake," we use the looping method to impose this restriction.


<div class="alert alert-block alert-warning" style="font-size: 20px;"> 
  <b> Part III: </b>Model Formulations
</div>

### **(a) Build the Algebraic Formulation**
_Formulate (algebraically) an optimization problem to make a daily diet plan for each child, with the goal of minimizing the total cost of the food while satisfying the general nutritional requirements and additional considerations stated above._

To restate the optimization problem based on the scenario described, we are seeking to design a weekly meal plan that provides the necessary nutrients for children living in poverty, at the lowest possible cost.

Here's the reformulated problem:

#### **Variables:**
- Xi: number of servings of food item i per child per day where each food item i corresponds to the food options in the dataset.
- Ci: cost per serving of food item i

#### **Objective Function:**
Minimizing the total cost of the food while satisfying the general nutritional requirements and additional considerations$$\min \sum_{i=1}^{n} c_i x_i$$ 

#### **Nutritional Constraints:**
$$1800 \le \sum_{i=1}^{n} Calories_i.x_i \le 2400$$
$$60 \le \sum_{i=1}^{n} Fat_i.x_i \le 95$$
$$1200 \le \sum_{i=1}^{n} Sodium_i.x_i \le 2200$$
$$240 \le \sum_{i=1}^{n} Carbs_i.x_i \le 400$$
$$30 \le \sum_{i=1}^{n} Fiber_i.x_i \le 35$$
$$40 \le \sum_{i=1}^{n} Protein_i.x_i \le 55$$
$$2000 \le \sum_{i=1}^{n} Vitamin A_i.x_i \le 6000$$
$$45 \le \sum_{i=1}^{n} Vitamin C_i.x_i \le 1200$$
$$1300 \le \sum_{i=1}^{n} Calcium_i.x_i \le 3000$$
$$8 \le \sum_{i=1}^{n} Iron_i.x_i \le 40$$

#### **Additional Constraints:**
$$\text{for each i:  } Calories_i.x_i \le 0.30 *  \sum_{j=1}^{n} Calories_j.x_j$$
$$\text{for each i:  } Protein_i.x_i \le 0.30 *  \sum_{j=1}^{n} Protein_j.x_j$$
where j is the index representing all food items, used in the summation.

#### **Non-negativity constraints:**
$$x_i \ge 0 \text{ for all i}$$

### **(b) Generate the Code Using Gurobi**

_Solve your formulation from part a) using Python/Gurobi to come up with a recommendation for the daily meal plan. Indicate the daily cost and the amount of each ingredient._

In [29]:
import gurobipy as gp
from gurobipy import GRB
import pandas as pd

# Loading the food data and nutritional requirements from Excel
file_path = 'food_data.xlsx'
food_data = pd.read_excel(file_path, sheet_name=1)

# Nutritional requirements (min and max) from the first sheet
nutritional_req = pd.read_excel(file_path, sheet_name=0)

# Creating the Gurobi model
model = gp.Model("Meal Plan Optimization")

# Creating decision variables for each food item (number of servings)
servings = model.addVars(food_data['Food'], lb=0, name="Servings")

# Setting objective function: minimize the total cost
model.setObjective(gp.quicksum(servings[food] * food_data.loc[food_data['Food'] == food, 'Cost ($/serving)'].values[0] 
                               for food in food_data['Food']), GRB.MINIMIZE)

# Adding constraints for each nutrient (Calories, Fat, Sodium, etc.)
for nutrient in nutritional_req['Nutrient']:
    min_val = nutritional_req.loc[nutritional_req['Nutrient'] == nutrient, 'Min'].values[0]
    max_val = nutritional_req.loc[nutritional_req['Nutrient'] == nutrient, 'Max'].values[0]
    
    # Minimum constraint
    model.addConstr(gp.quicksum(servings[food] * food_data.loc[food_data['Food'] == food, nutrient].values[0]
                                for food in food_data['Food']) >= min_val, name=f"{nutrient}_min")
    
    # Maximum constraint
    model.addConstr(gp.quicksum(servings[food] * food_data.loc[food_data['Food'] == food, nutrient].values[0]
                                for food in food_data['Food']) <= max_val, name=f"{nutrient}_max")

# No single food item should make up more than 30% of total calories or protein intake
model.addConstrs((servings[food] * food_data.loc[food_data['Food'] == food, 'Calories (kcal)'].values[0] <= 
                  0.30 * gp.quicksum(servings[f] * food_data.loc[food_data['Food'] == f, 'Calories (kcal)'].values[0]
                                     for f in food_data['Food'])
                  for food in food_data['Food']), name="Calorie_Limit")

model.addConstrs((servings[food] * food_data.loc[food_data['Food'] == food, 'Protein (g)'].values[0] <= 
                  0.30 * gp.quicksum(servings[f] * food_data.loc[food_data['Food'] == f, 'Protein (g)'].values[0]
                                     for f in food_data['Food'])
                  for food in food_data['Food']), name="Protein_Limit")

# Optimizing the model
model.optimize()

# Printing the results
if model.status == GRB.OPTIMAL:
    print(f"Optimal total daily cost: ${model.objVal:.2f}")
    
    # Displaying servings of each food item
    for food in food_data['Food']:
        if servings[food].x > 0:
            print(f"{food}: {servings[food].x:.2f} servings")
    
    # Calculating and displaying the total sum of each nutrient
    print("\nTotal Nutrients:")
    for nutrient in nutritional_req['Nutrient']:
        total_nutrient = gp.quicksum(servings[food].x * food_data.loc[food_data['Food'] == food, nutrient].values[0]
                                     for food in food_data['Food']).getValue()
        print(f"Total {nutrient}: {total_nutrient:.2f}")
else:
    print("No optimal solution found.")

Gurobi Optimizer version 11.0.3 build v11.0.3rc0 (win64 - Windows 11.0 (22631.2))

CPU model: 13th Gen Intel(R) Core(TM) i7-13700H, instruction set [SSE2|AVX|AVX2]
Thread count: 14 physical cores, 20 logical processors, using up to 20 threads

Optimize a model with 80 rows, 30 columns and 2326 nonzeros
Model fingerprint: 0x97023c77
Coefficient statistics:
  Matrix range     [6e-02, 2e+04]
  Objective range  [2e-02, 8e-01]
  Bounds range     [0e+00, 0e+00]
  RHS range        [8e+00, 6e+03]
Presolve removed 10 rows and 0 columns
Presolve time: 0.01s
Presolved: 70 rows, 40 columns, 2073 nonzeros

Iteration    Objective       Primal Inf.    Dual Inf.      Time
       0    0.0000000e+00   2.294375e+02   0.000000e+00      0s
      20    2.4310277e+00   0.000000e+00   0.000000e+00      0s

Solved in 20 iterations and 0.02 seconds (0.00 work units)
Optimal objective  2.431027707e+00
Optimal total daily cost: $2.43
Skim Milk: 1.00 servings
2% Lowfat Milk: 2.04 servings
Pork: 0.57 servings
Banan

<div class="alert alert-block alert-warning" style="font-size: 20px;"> 
  <b> Part IV: </b>Model Results
</div>



Based on the calculation results, we have the following meal plan:


\begin{array}{c}
\textbf{\Large Daily Meal Plan} \\
\end{array}

\begin{array}{|l|c|}
\hline
\textbf{Food Item} & \textbf{Servings} \\
\hline
\text{Skim Milk} & 1.00 \\
\text{2\% Lowfat Milk} & 2.04 \\
\text{Pork} & 0.57 \\
\text{Banana} & 0.09 \\
\text{Oranges} & 4.54 \\
\text{Potatoes, Baked} & 3.15 \\
\text{Spaghetti W/ Sauce} & 0.61 \\
\text{White Rice} & 0.16 \\
\hline
\end{array}


\begin{array}{c}
\textbf{\Large Total Daily Nutrition} \\
\end{array}

\begin{array}{|l|c|}
\hline
\textbf{Nutrient} & \textbf{Total Amount} \\
\hline
\text{Calories (kcal)} & 1800.00 \\
\text{Fat (g)} & 60.00 \\
\text{Sodium (mg)} & 1200.00 \\
\text{Carbs (g)} & 269.56 \\
\text{Fiber (g)} & 35.00 \\
\text{Protein (g)} & 55.00 \\
\text{Vitamin A (IU)} & 4622.72 \\
\text{Vitamin C (mg)} & 390.70 \\
\text{Calcium (mg)} & 1300.00 \\
\text{Iron (mg)} & 17.21 \\
\text{Total Daily Cost} & \$2.43 \\
\hline
\end{array}


\begin{array}{c}
\textbf{Optimal total daily cost: \$2.43} \\
\end{array}



<div class="alert alert-block alert-warning" style="font-size: 20px;"> 
  <b> Part V: </b>Discussions / Recommendations / Conclusions
</div>


### **(c) Reasonability and Recommendations**

_Discuss whether you think the recommendation from part b) sounds reasonable for a single day of food intake for a child.  If not, suggest some changes you would make to the model to make the recommendation for the day more realistic.  Implement your change and discuss the new solution._

#### **Issues:**
Based on the meal plan above, there are a few key issues with the current menu:
1. **Lack of Vegetables:** The meal plan does not include any vegetables, which contradicts the World Health Organization's (WHO) recommendations that adolescents should consume at least 400 grams or five portions of fruits and vegetables daily to mitigate the risk of non-communicable diseases (NCDs) and to ensure adequate fiber intake. 
(World Health Organization (WHO) )
_https://www.who.int/publications/i/item/9789241513708._
2. **Dairy Homogenization:** The inclusion of both skim milk and 2% lowfat milk may not provide the variety needed in terms of nutrient profile and could lead to an over-reliance on dairy for certain nutrients.
3. **High Fruit Servings with Limited Variety:** While the plan includes a substantial amount of oranges (4.54 servings), there is a lack of variety. Relying on a single fruit does not provide the broad spectrum of nutrients found in different fruits, and it may not meet the recommended daily intake when considering the total grams recommended by WHO.
4. **Excessive Carbohydrate Intake from Starches:** The plan includes three types of starchy foods (baked potatoes, spaghetti with sauce, and white rice), which could lead to excessive carbohydrate intake and potentially higher caloric consumption than necessary, especially if not balanced with adequate physical activity.
5. **Low Meat Serving Size:** With only 0.57 servings of pork, the meal plan may not provide enough protein to meet the daily requirements, which is essential for growth and maintenance in children.

#### **Recommendations:**
1. **Introduce a Variety of Vegetables:** To align with WHO's recommendations, incorporate a variety of vegetables into the meal plan. This could include a mix of leafy greens, root vegetables, and other colorful vegetables to provide a wide range of nutrients and fiber.
2. **Balance Macronutrients:** Adjust the proportions of carbohydrates, proteins, and fats to ensure a balanced meal that meets the nutritional needs without exceeding caloric intake.
3. **Increase Meat Sources:** Consider adding more servings of meat such as chicken, turkey and so on.
4. **Avoid Dairy Homogenization:** While dairy can be part of a healthy diet, ensure it is consumed in moderation alongside other food groups to maintain a balanced diet.
By addressing these issues, the meal plan can be improved to better support the health and nutritional needs of the children.

#### **Optimization Plan:**
1.	We have set an upper limit on each food item: "servings cannot exceed 2."
2.	We have divided the foods into the following groups:
>> **Vegetables:** Broccoli, Carrots (Raw), Corn, Lettuce (Iceberg, Raw), Peppers (Sweet, Raw), Potatoes (Baked), Tomato (Red, Ripe, Raw)

>> **Fruits:** Apple (Raw, w/Skin), Banana, Grapes, Kiwifruit (Raw, Fresh), Oranges

>> **Grains/Breads:** Bagels, Wheat Bread, White Bread, Oatmeal, Couscous, White Rice

>> **Proteins:** Roasted Chicken, Turkey, Beef, Pork, White Tuna in Water, Poached Eggs, Scrambled Eggs, Tofu

>> **Dairy:** 2% Lowfat Milk, Skim Milk

>> **Other:** Spaghetti W/ Sauce, Macaroni (cooked)

> Each group must contain at least one food item (in case of vegetables atleast 2)

3.	Require at least 10 different food items to be included in the meal plan


#### **Creating the Updated Model**

In [3]:
import gurobipy as gp
from gurobipy import GRB
import pandas as pd

# Loading the food data and nutritional requirements from Excel
file_path = 'food_data.xlsx'  
food_data = pd.read_excel(file_path, sheet_name=1)

# Nutritional requirements (min and max) from the first sheet
nutritional_req = pd.read_excel(file_path, sheet_name=0)

# Creating the Gurobi model
model = gp.Model("Meal Plan Optimization")

# Creating decision variables for each food item (number of servings)
servings = model.addVars(food_data['Food'], lb=0, name="Servings")

# Creating binary variables to indicate whether a food is included in the diet
include_food = model.addVars(food_data['Food'], vtype=GRB.BINARY, name="IncludeFood")

# Setting an upper limit on each food item: servings cannot exceed 2
max_servings = 2
model.addConstrs((servings[food] <= max_servings for food in food_data['Food']), name="Max_Servings_Per_Food")

# Linking servings and binary variables (if servings > 0, then include_food = 1)
model.addConstrs((servings[food] >= 0.1 * include_food[food] for food in food_data['Food']), name="Link_Serving_Include")

# Setting objective function: minimize the total cost
model.setObjective(
    gp.quicksum(servings[food] * food_data.loc[food_data['Food'] == food, 'Cost ($/serving)'].values[0]
               for food in food_data['Food']),
    GRB.MINIMIZE
)

# Adding constraints for each nutrient (Calories, Fat, Sodium, etc.)
for nutrient in nutritional_req['Nutrient']:
    min_val = nutritional_req.loc[nutritional_req['Nutrient'] == nutrient, 'Min'].values[0]
    max_val = nutritional_req.loc[nutritional_req['Nutrient'] == nutrient, 'Max'].values[0]
    
    # Minimum constraint
    model.addConstr(gp.quicksum(servings[food] * food_data.loc[food_data['Food'] == food, nutrient].values[0]
                                for food in food_data['Food']) >= min_val, name=f"{nutrient}_min")
    
    # Maximum constraint
    model.addConstr(gp.quicksum(servings[food] * food_data.loc[food_data['Food'] == food, nutrient].values[0]
                                for food in food_data['Food']) <= max_val, name=f"{nutrient}_max")

# Ensuring that no single food item exceeds 30% of total calories or protein intake
model.addConstrs((servings[food] * food_data.loc[food_data['Food'] == food, 'Calories (kcal)'].values[0] <= 
                  0.30 * gp.quicksum(servings[f] * food_data.loc[food_data['Food'] == f, 'Calories (kcal)'].values[0]
                                     for f in food_data['Food'])
                  for food in food_data['Food']), name="Calorie_Limit")

model.addConstrs((servings[food] * food_data.loc[food_data['Food'] == food, 'Protein (g)'].values[0] <= 
                  0.30 * gp.quicksum(servings[f] * food_data.loc[food_data['Food'] == f, 'Protein (g)'].values[0]
                                     for f in food_data['Food'])
                  for food in food_data['Food']), name="Protein_Limit")

# Grouping food items and ensure at least one from each group is included
vegetables = ['Broccoli', 'Carrots, Raw', 'Corn', 'Lettuce, Iceberg, Raw', 'Peppers, Sweet, Raw', 
              'Potatoes, Baked', 'Tomato, Red, Ripe, Raw']
fruits = ['Apple, Raw, w/Skin', 'Banana', 'Grapes', 'Kiwifruit, Raw, Fresh', 'Oranges']
grains_breads = ['Bagels', 'Wheat Bread', 'White Bread', 'Oatmeal', 'Couscous', 'White Rice']
proteins = ['Roasted Chicken', 'Turkey', 'Beef', 'Pork', 'White Tuna in Water', 'Poached Eggs', 'Scrambled Eggs', 'Tofu']
dairy = ['2% Lowfat Milk', 'Skim Milk']
other = ['Spaghetti W/ Sauce', 'Macaroni, cooked']

# Adding constraints to include at least one food from each group
model.addConstr(gp.quicksum(include_food[food] for food in vegetables if food in food_data['Food'].values) >= 2, name="Min_Vegetables")
model.addConstr(gp.quicksum(include_food[food] for food in fruits) >= 1, name="Min_Fruits")
model.addConstr(gp.quicksum(include_food[food] for food in grains_breads) >= 1, name="Min_Grains_Breads")
model.addConstr(gp.quicksum(include_food[food] for food in proteins) >= 1, name="Min_Proteins")
model.addConstr(gp.quicksum(include_food[food] for food in dairy) >= 1, name="Min_Dairy")
model.addConstr(gp.quicksum(include_food[food] for food in other) >= 1, name="Min_Other")

# Ensuring at least 10 different food items are included in the meal plan
model.addConstr(gp.quicksum(include_food[food] for food in food_data['Food']) >= 10, name="Min_Variety")

# Optimizing the model
model.optimize()

# Printing the results
if model.status == GRB.OPTIMAL:
    print(f"Optimal total daily cost: ${model.objVal:.2f}")
    
    # Displaying servings of each food item
    for food in food_data['Food']:
        if servings[food].x > 0:
            print(f"{food}: {servings[food].x:.2f} servings")
    
    # Calculating and displaying the total sum of each nutrient
    print("\nTotal Nutrients:")
    for nutrient in nutritional_req['Nutrient']:
        total_nutrient = gp.quicksum(servings[food].x * food_data.loc[food_data['Food'] == food, nutrient].values[0]
                                     for food in food_data['Food']).getValue()
        print(f"Total {nutrient}: {total_nutrient:.2f}")
else:
    print("No optimal solution found.")


Gurobi Optimizer version 11.0.3 build v11.0.3rc0 (win64 - Windows 11+.0 (26100.2))

CPU model: 12th Gen Intel(R) Core(TM) i7-12650H, instruction set [SSE2|AVX|AVX2]
Thread count: 10 physical cores, 16 logical processors, using up to 16 threads

Optimize a model with 147 rows, 60 columns and 2474 nonzeros
Model fingerprint: 0x8d9ec198
Variable types: 30 continuous, 30 integer (30 binary)
Coefficient statistics:
  Matrix range     [6e-02, 2e+04]
  Objective range  [2e-02, 8e-01]
  Bounds range     [1e+00, 1e+00]
  RHS range        [1e+00, 6e+03]
Presolve removed 32 rows and 0 columns
Presolve time: 0.00s
Presolved: 115 rows, 60 columns, 2407 nonzeros
Variable types: 30 continuous, 30 integer (30 binary)

Root relaxation: objective 2.513664e+00, 53 iterations, 0.00 seconds (0.00 work units)

    Nodes    |    Current Node    |     Objective Bounds      |     Work
 Expl Unexpl |  Obj  Depth IntInf | Incumbent    BestBd   Gap | It/Node Time

*    0     0               0       2.5136637    2

### **Optimized Model Results:**

Based on the updated results, we have the following meal plan:


\begin{array}{c}
\textbf{\Large Daily Meal Plan} \\
\end{array}

\begin{array}{|l|c|}
\hline
\textbf{Food Item} & \textbf{Servings} \\
\hline
\text{Carrots, Raw} & 0.10 \\
\text{Potatoes, Baked} & 2.00 \\
\text{Spaghetti W/ Sauce} & 0.57 \\
\text{Apple, Raw, w/Skin} & 0.52 \\
\text{Banana} & 2.00 \\
\text{Oranges} & 2.00 \\
\text{2\% Lowfat Milk} & 2.00 \\
\text{Skim Milk} & 1.49 \\
\text{White Rice} & 0.37 \\
\text{Pork} & 0.66 \\
\hline
\end{array}




\begin{array}{c}
\textbf{\Large Total Daily Nutrition} \\
\end{array}

\begin{array}{|l|c|}
\hline
\textbf{Nutrient} & \textbf{Total Amount} \\
\hline
\text{Total Calories (kcal)} & 1800.00 \\
\text{Total Fat (g)} & 66.42 \\
\text{Total Sodium (mg)} & 1200.00 \\
\text{Total Carbs (g)} & 250.39 \\
\text{Total Fiber (g)} & 35.00 \\
\text{Total Protein (g)} & 55.00 \\
\text{Total Vitamin A (IU)} & 5813.08 \\
\text{Total Vitamin C (mg)} & 220.20 \\
\text{Total Calcium (mg)} & 1300.00 \\
\text{Total Iron (mg)} & 14.59 \\
\hline
\end{array}

\begin{array}{c}
\textbf{Optimal total daily cost: \$2.51} \\
\end{array}


The new constraints will likely lead to a slightly higher cost but will produce a more practical and balanced meal plan. The maximum serving size constraint prevents overloading one food item, and the variety constraint ensures the child gets a balanced intake from different sources.

### **Optimized Model Results Analysis**

By introducing food frequency constraints and varying meal components, we ensure that children receive a diverse and balanced diet throughout the week. The optimized model allows for:

#### **Nutritional Adequacy**
> The meal plan meets the daily nutritional needs while adhering to dietary guidelines. It provides a balanced intake of macronutrients such as carbohydrates, fats, and proteins, all within the recommended ranges. For instance, the plan offers 1800 kcal, which is within the 1800-2400 kcal range, and 55g of protein, aligning with the 40-55g recommendation. Additionally, micronutrients like Vitamin A, Vitamin C, Calcium, and Iron are all met, indicating that the plan is nutritionally adequate.

#### **Meal Variety and Nutritional Balance**
> The plan introduces a variety of foods across different food groups, which helps maintain interest and enjoyment while ensuring a nutritionally balanced diet. The inclusion of fruits like bananas and oranges, vegetables with the raw apple, and a mix of protein sources from milk, pork, and bread provides a diverse range of nutrients. Combining meal variety with nutritional balance is crucial for overall health and to prevent dietary monotony.

#### **Cost Control**
> The overall cost of the meal plan is minimized at $2.51 per day, which is an affordable daily expenditure for nutritious meals. This cost control is achieved without compromising on the quality or variety of food, indicating efficient menu planning and cost management. Maintaining cost efficiency is essential for the sustainability of such a meal plan, especially when catering to a large number of children.

#### **Practicality**
> The meal plan is practical in terms of serving sizes and meal compositions suitable for children. Portion sizes are reasonable and align with dietary recommendations to prevent overeating or nutrient deficiencies. The meal compositions are also practical, considering the ease of preparation and the likelihood of children consuming the offered foods. For example, baked potatoes, white rice, and spaghetti are familiar and easily accessible items that are typically well-received by children.
In summary, the meal plan demonstrates nutritional adequacy by meeting daily requirements, offers variety to ensure interest and enjoyment, maintains cost control for affordability, and is practical for implementation in a real-world setting suitable for children. These factors contribute to the overall effectiveness of the meal plan in promoting health and well-being.


<div class="alert alert-block alert-warning" style="font-size: 20px;"> 
  <b> Part VI: </b>Weekly Meal Plan
</div>

### **(d) Creating a 7 Day Dietary Plan**

_Whereas part (c) asked you to consider whether the meal plan for the one day is sensible or not and how you might get to a reasonable plan for even one day.  However, Principal Harper also knows that kids don’t like to eat the same exact thing every day, so for part (d) she asks you to come up with a reasonable dietary plan for each day of an entire week.  Present a 7-day dietary plan, and indicate any model changes that went into constructing each day’s plan._

### **Explanation of the Code**

To answer the part (d), we created a 7-day meal plan by ensuring that nutritional requirements are met while minimizing costs and maintaining variety throughout the week. The solution was developed using Gurobi Optimizer and a model that incorporates dietary constraints for each day of the week.

#### **Key Elements of the Code:**

**Data Inputs:**
- We used a dataset of food items from an Excel file, which includes the cost and nutritional content of each item (e.g., calories, protein, fat, vitamins, etc.).
- The code loads two sheets: the first contains food data (with associated nutrients and costs), and the second contains nutritional requirements (e.g., minimum and maximum daily intake values for various nutrients).

**Food Groups:**
The foods were categorized into the following groups:
- **Vegetables:** Broccoli, Carrots, Corn, etc.
- **Fruits:** Apples, Bananas, Grapes, etc.
- **Grains/Breads:** Wheat Bread, White Bread, Oatmeal, etc.
- **Proteins:** Chicken, Beef, Pork, Eggs, etc.
- **Dairy:** 2% Lowfat Milk, Skim Milk
- **Other:** Spaghetti with Sauce, Macaroni

**Daily Constraints:**
Each day had a specific set of constraints on how many servings from each food group should be included:
- Example for Day 1: At least 2 vegetables, 2 proteins, 1 grain/bread, 1 fruit, 1 dairy, and 1 "other" item.
- These constraints varied day by day to ensure variety in the meal plans.

**Model Optimization:**
- The code creates a Gurobi model that defines the objective to minimize the total weekly cost while ensuring that all the nutritional requirements are met (e.g., calories, fat, protein, vitamins, etc.).
- Slack and excess variables are included to handle soft constraints on nutritional intake (e.g., slightly exceeding the maximum or falling below the minimum requirements adds a penalty to the cost).

**Additional Constraints:**
- **Variety:** Each meal plan must include at least 6 different food items per day.
- **Food Group Proportions:** No single food should contribute more than 30% of total calories or protein intake.
- **Max Servings:** A limit of 3 servings per food was imposed to avoid overreliance on a single item.

**Daily Optimization:**
For each day, the model optimizes the meal plan by selecting the optimal combination of foods that meet the given constraints and minimizes the cost. If an optimal solution is found, the code prints the total cost, the food items chosen, the number of servings, and the total intake of each nutrient.


In [2]:
import gurobipy as gp
from gurobipy import GRB
import pandas as pd

# Loading the food data and nutritional requirements from Excel
file_path = 'food_data.xlsx' 
food_data = pd.read_excel(file_path, sheet_name=1)
nutritional_req = pd.read_excel(file_path, sheet_name=0)

# Defining various food groups
vegetables = ['Broccoli', 'Carrots, Raw', 'Corn', 'Lettuce, Iceberg, Raw', 'Peppers, Sweet, Raw', 'Potatoes, Baked', 'Tomato, Red, Ripe, Raw']
fruits = ['Apple, Raw, w/Skin', 'Banana', 'Grapes', 'Kiwifruit, Raw, Fresh', 'Oranges']
grains_breads = ['Bagels', 'Wheat Bread', 'White Bread', 'Oatmeal', 'Couscous', 'White Rice', 'Macaroni, cooked']
proteins = ['Roasted Chicken', 'Turkey', 'Beef', 'Pork', 'White Tuna in Water', 'Poached Eggs', 'Scrambled Eggs', 'Tofu']
dairy = ['2% Lowfat Milk', 'Skim Milk']
other = ['Spaghetti W/ Sauce', 'Macaroni, cooked']

# Days of the week (weekly diet planning)
DAYS = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday']

# Daily constraints for each day (for variety within the day)
daily_constraints = [
    {"vegetables": 2, "proteins": 2, "grains_breads": 1, "fruits": 1, "dairy": 1, "other": 1},
    {"vegetables": 3, "proteins": 1, "grains_breads": 2, "fruits": 0, "dairy": 0, "other": 0},
    {"vegetables": 1, "proteins": 1, "grains_breads": 1, "fruits": 1, "dairy": 1, "other": 1},
    {"vegetables": 2, "proteins": 2, "grains_breads": 1, "fruits": 0, "dairy": 0, "other": 0},
    {"vegetables": 2, "proteins": 1, "grains_breads": 1, "fruits": 1, "dairy": 0, "other": 0},
    {"vegetables": 3, "proteins": 2, "grains_breads": 1, "fruits": 0, "dairy": 0, "other": 0},
    {"vegetables": 3, "proteins": 2, "grains_breads": 2, "fruits": 2, "dairy": 0, "other": 1}
]

# Initializing the model
model = gp.Model("7-Day Dietary Plan with Weekly Variety")

# Decision variables: servings for each food on each day
servings = model.addVars(food_data['Food'], DAYS, lb=0, name="Servings")

# Binary variables to track if a food is included on a particular day
include_food = model.addVars(food_data['Food'], DAYS, vtype=GRB.BINARY, name="IncludeFood")

# Slack and excess variables for soft constraints
slack_vars = model.addVars(nutritional_req['Nutrient'], DAYS, lb=0, name="Slack")
excess_vars = model.addVars(nutritional_req['Nutrient'], DAYS, lb=0, name="Excess")

# Setting objective: minimize total cost plus penalties for slack and excess
model.setObjective(
    gp.quicksum(
        servings[food, day] * food_data.loc[food_data['Food'] == food, 'Cost ($/serving)'].values[0]
        for food in food_data['Food'] for day in DAYS
    ) +
    gp.quicksum(slack_vars[nutrient, day] * 10 for nutrient in nutritional_req['Nutrient'] for day in DAYS) +
    gp.quicksum(excess_vars[nutrient, day] * 0.2 for nutrient in nutritional_req['Nutrient'] for day in DAYS),
    GRB.MINIMIZE
)

# Adding constraints for each day
for idx, day in enumerate(DAYS):
    constraints = daily_constraints[idx]
    
    # Nutritional constraints
    for nutrient in nutritional_req['Nutrient']:
        min_val = nutritional_req.loc[nutritional_req['Nutrient'] == nutrient, 'Min'].values[0]
        max_val = nutritional_req.loc[nutritional_req['Nutrient'] == nutrient, 'Max'].values[0]
        
        # Minimum constraint with slack
        model.addConstr(
            gp.quicksum(servings[food, day] * food_data.loc[food_data['Food'] == food, nutrient].values[0]
                        for food in food_data['Food']) + slack_vars[nutrient, day] >= min_val,
            name=f"{nutrient}_min_{day}"
        )
        
        # Maximum constraint with excess
        model.addConstr(
            gp.quicksum(servings[food, day] * food_data.loc[food_data['Food'] == food, nutrient].values[0]
                        for food in food_data['Food']) - excess_vars[nutrient, day] <= max_val,
            name=f"{nutrient}_max_{day}"
        )
    
    # Maximum servings per food per day
    max_servings = 3
    model.addConstrs(
        (servings[food, day] <= max_servings * include_food[food, day] for food in food_data['Food']),
        name=f"Max_Servings_Per_Food_{day}"
    )
    
    # Linking serving size with binary variable
    model.addConstrs(
        (servings[food, day] >= 0.1 * include_food[food, day] for food in food_data['Food']),
        name=f"Link_Serving_Include_{day}"
    )
    
    # Minimum number of food items for variety within the day
    model.addConstr(
        gp.quicksum(include_food[food, day] for food in food_data['Food']) >= 6,
        name=f"Min_Variety_{day}"
    )
    
    # No single food should exceed 30% of total calories or protein
    total_calories = gp.quicksum(servings[food, day] * food_data.loc[food_data['Food'] == food, 'Calories (kcal)'].values[0]
                                 for food in food_data['Food'])
    total_protein = gp.quicksum(servings[food, day] * food_data.loc[food_data['Food'] == food, 'Protein (g)'].values[0]
                                for food in food_data['Food'])
    
    for food in food_data['Food']:
        model.addConstr(
            servings[food, day] * food_data.loc[food_data['Food'] == food, 'Calories (kcal)'].values[0] <= 0.3 * total_calories,
            name=f"Cal30_{food}_{day}"
        )
        model.addConstr(
            servings[food, day] * food_data.loc[food_data['Food'] == food, 'Protein (g)'].values[0] <= 0.3 * total_protein,
            name=f"Protein30_{food}_{day}"
        )
    
    # Applying daily food group constraints
    model.addConstr(
    gp.quicksum(include_food[food, day] for food in vegetables if food in food_data['Food'].values) >= constraints["vegetables"],
    name=f"Min_Vegetables_{day}"
)
    model.addConstr(
        gp.quicksum(include_food[food, day] for food in proteins) >= constraints["proteins"],
        name=f"Min_Proteins_{day}"
    )
    model.addConstr(
        gp.quicksum(include_food[food, day] for food in grains_breads) >= constraints["grains_breads"],
        name=f"Min_Grains_Breads_{day}"
    )
    model.addConstr(
        gp.quicksum(include_food[food, day] for food in fruits) >= constraints["fruits"],
        name=f"Min_Fruits_{day}"
    )
    model.addConstr(
        gp.quicksum(include_food[food, day] for food in dairy) >= constraints["dairy"],
        name=f"Min_Dairy_{day}"
    )
    model.addConstr(
        gp.quicksum(include_food[food, day] for food in other) >= constraints["other"],
        name=f"Min_Other_{day}"
    )

# Weekly variety constraints
max_days_per_food = 4 # A food can appear at most 4 days in the week
model.addConstrs(
    (gp.quicksum(include_food[food, day] for day in DAYS) <= max_days_per_food for food in food_data['Food']),
    name="Max_Days_Per_Food"
)

# Optimizing the model
model.optimize()

# Checking if the model has an optimal solution
if model.status == GRB.OPTIMAL:
    print(f"\nOptimal total weekly cost: ${model.objVal:.2f}\n")
    for day in DAYS:
        print(f"--- {day} ---")
        day_cost = 0
        for food in food_data['Food']:
            qty = servings[food, day].X
            if qty > 0.01:  # Only display foods that are actually used
                cost = qty * food_data.loc[food_data['Food'] == food, 'Cost ($/serving)'].values[0]
                day_cost += cost
                print(f"{food}: {qty:.2f} servings (${cost:.2f})")
        print(f"Total cost for {day}: ${day_cost:.2f}\n")
else:
    print("No optimal solution found.")

Set parameter Username
Academic license - for non-commercial use only - expires 2025-09-06
Gurobi Optimizer version 11.0.3 build v11.0.3rc0 (win64 - Windows 11+.0 (26100.2))

CPU model: 12th Gen Intel(R) Core(TM) i7-12650H, instruction set [SSE2|AVX|AVX2]
Thread count: 10 physical cores, 16 logical processors, using up to 16 threads

Optimize a model with 1059 rows, 560 columns and 17885 nonzeros
Model fingerprint: 0x86737207
Variable types: 350 continuous, 210 integer (210 binary)
Coefficient statistics:
  Matrix range     [6e-02, 2e+04]
  Objective range  [2e-02, 1e+01]
  Bounds range     [1e+00, 1e+00]
  RHS range        [1e+00, 6e+03]
Found heuristic solution: objective 303356.25955
Presolve added 0 rows and 70 columns
Presolve removed 82 rows and 0 columns
Presolve time: 0.03s
Presolved: 977 rows, 630 columns, 16143 nonzeros
Variable types: 420 continuous, 210 integer (210 binary)

Root relaxation: objective 1.896982e+01, 809 iterations, 0.03 seconds (0.03 work units)

    Nodes  

### **Results and Explanations by Day:**

\begin{array}{c}
\textbf{\Large 7-Day Meal Plan} \\
\end{array}

\begin{array}{|l|c|c|}
\hline
\textbf{Day} & \textbf{Food Item} & \textbf{Servings (Cost)} \\
\hline
\text{Monday} & \text{Carrots, Raw} & 0.11 \, (0.01) \\
& \text{Potatoes, Baked} & 3.00 \, (0.18) \\
& \text{Tofu} & 2.05 \, (0.64) \\
& \text{Spaghetti W/ Sauce} & 0.65 \, (0.50) \\
& \text{Apple, Raw, w/Skin} & 1.24 \, (0.30) \\
& \text{Oranges} & 3.00 \, (0.45) \\
& \text{Wheat Bread} & 0.10 \, (0.01) \\
& \text{Skim Milk} & 2.41 \, (0.31) \\
& \text{Pork} & 0.52 \, (0.43) \\
\textbf{Total Cost:} & & \textbf{\$2.82} \\
\hline

\text{Tuesday} & \text{Broccoli} & 0.50 \, (0.08) \\
& \text{Corn} & 0.10 \, (0.02) \\
& \text{Potatoes, Baked} & 2.88 \, (0.17) \\
& \text{Oranges} & 3.00 \, (0.45) \\
& \text{White Bread} & 3.00 \, (0.18) \\
& \text{2\% Lowfat Milk} & 2.72 \, (0.63) \\
& \text{Scrambled Eggs} & 2.20 \, (0.24) \\
& \text{Beef} & 2.07 \, (0.56) \\
& \text{White Rice} & 0.39 \, (0.03) \\
\textbf{Total Cost:} & & \textbf{\$2.36} \\
\hline

\text{Wednesday} & \text{Carrots, Raw} & 0.10 \, (0.01) \\
& \text{Spaghetti W/ Sauce} & 0.23 \, (0.18) \\
& \text{Apple, Raw, w/Skin} & 2.99 \, (0.72) \\
& \text{Banana} & 3.00 \, (0.45) \\
& \text{Wheat Bread} & 3.00 \, (0.15) \\
& \text{2\% Lowfat Milk} & 2.04 \, (0.47) \\
& \text{Skim Milk} & 1.82 \, (0.24) \\
& \text{White Rice} & 0.41 \, (0.03) \\
& \text{Pork} & 0.73 \, (0.59) \\
\textbf{Total Cost:} & & \textbf{\$2.83} \\
\hline

\text{Thursday} & \text{Broccoli} & 0.69 \, (0.11) \\
& \text{Corn} & 0.10 \, (0.02) \\
& \text{Tofu} & 1.90 \, (0.59) \\
& \text{Apple, Raw, w/Skin} & 2.99 \, (0.72) \\
& \text{Banana} & 3.00 \, (0.45) \\
& \text{Wheat Bread} & 2.94 \, (0.15) \\
& \text{White Bread} & 3.00 \, (0.18) \\
& \text{Skim Milk} & 2.53 \, (0.33) \\
& \text{Pork} & 0.58 \, (0.47) \\
\textbf{Total Cost:} & & \textbf{\$3.01} \\
\hline

\text{Friday} & \text{Broccoli} & 0.50 \, (0.08) \\
& \text{Potatoes, Baked} & 2.91 \, (0.17) \\
& \text{Oranges} & 3.00 \, (0.45) \\
& \text{White Bread} & 3.00 \, (0.18) \\
& \text{2\% Lowfat Milk} & 2.72 \, (0.62) \\
& \text{Scrambled Eggs} & 2.20 \, (0.24) \\
& \text{Beef} & 2.08 \, (0.56) \\
& \text{White Rice} & 0.40 \, (0.03) \\
\textbf{Total Cost:} & & \textbf{\$2.34} \\
\hline

\text{Saturday} & \text{Carrots, Raw} & 0.10 \, (0.01) \\
& \text{Corn} & 0.10 \, (0.02) \\
& \text{Potatoes, Baked} & 3.00 \, (0.18) \\
& \text{Tofu} & 2.42 \, (0.75) \\
& \text{Spaghetti W/ Sauce} & 0.83 \, (0.65) \\
& \text{Banana} & 1.30 \, (0.19) \\
& \text{2\% Lowfat Milk} & 2.81 \, (0.65) \\
& \text{Scrambled Eggs} & 0.32 \, (0.03) \\
& \text{Beef} & 1.54 \, (0.41) \\
& \text{White Rice} & 0.38 \, (0.03) \\
\textbf{Total Cost:} & & \textbf{\$2.92} \\
\hline

\text{Sunday} & \text{Broccoli} & 0.54 \, (0.09) \\
& \text{Corn} & 0.10 \, (0.02) \\
& \text{Peppers, Sweet, Raw} & 0.10 \, (0.05) \\
& \text{Tofu} & 1.31 \, (0.41) \\
& \text{Spaghetti W/ Sauce} & 0.10 \, (0.08) \\
& \text{Apple, Raw, w/Skin} & 0.94 \, (0.22) \\
& \text{Banana} & 3.00 \, (0.45) \\
& \text{Oranges} & 3.00 \, (0.45) \\
& \text{Wheat Bread} & 2.25 \, (0.11) \\
& \text{White Bread} & 3.00 \, (0.18) \\
& \text{Skim Milk} & 2.37 \, (0.31) \\
& \text{Pork} & 0.68 \, (0.55) \\
\textbf{Total Cost:} & & \textbf{\$2.91} \\
\hline
\end{array}

\begin{array}{c}
\textbf{Optimal total weekly cost: \$38.60} \\
\end{array}



**Day 1:**
- **Constraints:** 2 vegetables, 2 proteins, 1 grain/bread, 1 fruit, 1 dairy, 1 other item.
- **Optimal Daily Cost:** $2.82
- **Selected Foods and Servings:**
  2% Lowfat Milk (2.41 servings), Skim Milk (2.41 servings), Potatoes (3.00 servings), Tofu (2.05 servings), etc.
  
**Day 2:**
- **Constraints:** 3 vegetables, 1 protein, 2 grains/breads, 0 fruits, 0 dairy, 0 other items.
- **Optimal Daily Cost:** $2.36
- **Selected Foods and Servings:**
  Beef (2.07 servings), Potatoes (2.88 servings), White Bread (3.00 servings), etc.
  
**Day 3:**
- **Constraints:** 1 vegetable, 1 protein, 1 grain/bread, 1 fruit, 1 dairy, 1 other item.
- **Optimal Daily Cost:** $2.83
- **Selected Foods and Servings:**
  Skim Milk (1.82 servings), White Rice (0.41 servings), Pork (0.73 servings), Apple (2.99 servings), etc.

**Day 4:**
- **Constraints:** 2 vegetables, 2 proteins, 1 grain/bread, 0 fruits, 0 dairy, 0 other items.
- **Optimal Daily Cost:** $3.01
- **Selected Foods and Servings:**
  Pork (0.58 servings), Tofu (1.90 servings), Wheat Bread (2.94 servings), etc.
  
**Day 5:**
- **Constraints:** 2 vegetables, 1 protein, 1 grain/bread, 1 fruit, 0 dairy, 0 other items.
- **Optimal Daily Cost:** $2.34
- **Selected Foods and Servings:**
  Scrambled Eggs (2.20 servings), White Bread (3.00 servings), Oranges (3.00 servings), etc.

**Day 6:**
- **Constraints:** 3 vegetables, 2 proteins, 1 grain/bread, 0 fruits, 0 dairy, 0 other items.
- **Optimal Daily Cost:** $2.92
- **Selected Foods and Servings:**
  Skim Milk (2.81 servings), Tofu (2.42 servings), Potatoes (3.00 servings), etc.

**Day 7:**
- **Constraints:** 3 vegetables, 2 proteins, 2 grains/breads, 2 fruits, 0 dairy, 1 other item.
- **Optimal Daily Cost:** $2.91
- **Selected Foods and Servings:**
  Pork (0.68 servings), Broccoli (0.54 servings), Apple (0.94 servings), Banana (3.00 servings), etc.

### **Conclusion:**

This updated approach ensured that each day's meal plan:
- Minimizes costs (between $2.34 and $3.01 per day).
- Meets the nutritional requirements for calories, protein, fat, vitamins, and minerals.
- Varies food selections across the week to avoid monotony, incorporating different types of vegetables, proteins, grains, and other food groups.
- Maintains diversity by ensuring a minimum of 6 food items each day and limiting the number of days any food appears.
